light/beacon/binary_merkle.go

```diff
@@ -0,0 +1,555 @@
+		hash := (*mpr.values)[0]
+		*mpr.values = (*mpr.values)[1:]
+		return hash, true
+	}
+	return MerkleValue{}, false
+}
+
+// Reader creates a multiProofReader for the given proof; if subtrees != nil
+// then also attaches subtree readers at indices where the function returns a
+// non-nil reader.
+// Note that the reader can only be traversed once as the values slice is
+// sequentially consumed.
+func (mp MultiProof) Reader(subtrees func(uint64) ProofReader) *multiProofReader {
+	values := mp.Values
+	return &multiProofReader{format: mp.Format, values: &values, index: 1, subtrees: subtrees}
+}
+
+// Finished returns true if all values have been consumed by the traversal.
+// Should be checked after TraverseProof if received from an untrusted source in
+// order to prevent DoS attacks by excess proof values.
+func (mpr multiProofReader) Finished() bool {
+	return len(*mpr.values) == 0
+}
+
+// rootHash returns the root hash of the proven structure.
+func (mp MultiProof) rootHash() common.Hash {
+	reader := mp.Reader(nil)
+	hash, ok := TraverseProof(reader, nil)
+	if !ok || !reader.Finished() {
+		log.Error("MultiProof.rootHash: invalid proof format")
+	}
+	return hash
+}
+
+// multiProofWriter implements ProofWriter and creates a MultiProof with the
+// previously specified format. Also allows attaching further subtree writers at
+// certain indices.
+type multiProofWriter struct {
+	format   ProofFormat
+	values   *MerkleValues
+	index    uint64
+	subtrees func(uint64) ProofWriter
+}
+
+// NewMultiProofWriter creates a new multiproof writer with the specified format.
+// If subtrees != nil then further subtree writers are attached at indices where
+// the function returns a non-nil writer.
+// Note that the specified format should not include these attached subtrees;
+// they should be attached at leaf indices of the given format.
+// Also note that target can be nil in which case the nodes specified by the format
+// are traversed but not stored; subtree writers might still store tree data.
+func NewMultiProofWriter(format ProofFormat, target *MerkleValues, subtrees func(uint64) ProofWriter) multiProofWriter {
+	return multiProofWriter{format: format, values: target, index: 1, subtrees: subtrees}
+}
+
+// children implements ProofWriter
+func (mpw multiProofWriter) children() (left, right ProofWriter) {
+	if mpw.subtrees != nil {
+		if subtree := mpw.subtrees(mpw.index); subtree != nil {
+			return subtree.children()
+		}
+	}
+	lf, rf := mpw.format.children()
+	if lf == nil {
+		return nil, nil
+	}
+	return multiProofWriter{format: lf, values: mpw.values, index: mpw.index * 2, subtrees: mpw.subtrees},
+		multiProofWriter{format: rf, values: mpw.values, index: mpw.index*2 + 1, subtrees: mpw.subtrees}
+}
+
+// writeNode implements ProofWriter
+func (mpw multiProofWriter) writeNode(node MerkleValue) {
+	if mpw.values != nil {
+		if lf, _ := mpw.format.children(); lf == nil {
+			*mpw.values = append(*mpw.values, node)
```

Comment:
Why not define function with `*multiProofWriter` as the function receiver, so that you can directly append value into `values`.

Reply:
Even if we could modify our `multiProofWriter` instance, that would not help us. The slice needs to be kept as a reference because the child writers append to the same list as the tree is traversed. If we copied the slice and the children appended to the copied slice, the new elements would not appear in the parent instance. I added explanatory comment to the code.